### ISSUE TITLE
NOPUCKPDB incorrect/unspecified bond length for ring-closing bonds

### ISSUE THREAD

**Issue by Alicecomma** [drsalonsogil/montecarbo#1]
The α-D-glucoses generated by NOPUCKPDB have incorrect ring-closing bond lengths. To show the issue, I plot here the C-C and C-O bond lengths of some rings generated by MonteCarbo:

![image](https://user-images.githubusercontent.com/4662406/151840073-48843a24-e375-4472-8ac6-978eb97f484a.png)

As can be seen, C1-C2-C3-C4-C5-O5 retain correct bond lengths, but O5-C1 bond lengths are on average 1.341 Angstrom (although N-6-H-1-Oxy.var specifies dOC=1.408) and have pretty severe variation. The original paper notes the main limitation is for 'extreme multi-functionalization cases with voluminous groups', however this problem seems to happen for species as simple as α-D-glucose, and seem unrelated to effects like steric hindrance.

It seems the ring-closing bond length is never set. According to Cremer (DOI: 10.1021/j100377a017), the ring conformation of an N-ring is fixed given N-3 puckering coordinates (φ,θ,Qrad), N-3 bond angles (randomized in MonteCarbo) and N bond lengths (however, only N-1 bond lengths are specified by MonteCarbo). Note in the MonteCarbo paper:

"_In case the bond distances and angles are known, N − 3 endocyclic dihedral angles are necessary to construct and define a specific conformation for an N-membered ring. In this article's following point, this geometrical property will be used to establish a random selection of conformers._"

I.e the assumption that all bond distances and angles are known is not implemented, as at least one bond distance is missing.

<details>
<summary>
**Method**
</summary>

The input code was:

`testglucose
NOPUCKPDB
6
1
Oxy
H
OH
H
OH
OH
H
H
OH
CH2OH
H
24`

Calculations of the atom-atom distances were performed by pasting the output .pdb files in Google Sheets and then extracting atom coordinates as follows:

```
X1=index(split(regexextract(regexextract(Input!J6,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,1)
Y1=index(split(regexextract(regexextract(Input!J6,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,2)
Z1=index(split(regexextract(regexextract(Input!J6,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,3)
X2=index(split(regexextract(regexextract(Input!J7,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,1)
Y2=index(split(regexextract(regexextract(Input!J7,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,2)
Z2=index(split(regexextract(regexextract(Input!J7,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,3)
X3=index(split(regexextract(regexextract(Input!J8,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,1)
Y3=index(split(regexextract(regexextract(Input!J8,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,2)
Z3=index(split(regexextract(regexextract(Input!J8,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,3)
X4=index(split(regexextract(regexextract(Input!J9,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,1)
Y4=index(split(regexextract(regexextract(Input!J9,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,2)
Z4=index(split(regexextract(regexextract(Input!J9,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,3)
X5=index(split(regexextract(regexextract(Input!J10,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,1)
Y5=index(split(regexextract(regexextract(Input!J10,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,2)
Z5=index(split(regexextract(regexextract(Input!J10,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,3)
X6=index(split(regexextract(regexextract(Input!J11,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,1)
Y6=index(split(regexextract(regexextract(Input!J11,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,2)
Z6=index(split(regexextract(regexextract(Input!J11,"[-]?\d\.\d\d\d.+"),".+\d\.\d\d\d")," ",true,true),1,3)
```

And then taking distance:

```
d(C1-C2)=sqrt((X1-X2)^2+(Y1-Y2)^2+(Z1-Z2)^2)
d(C2-C3)=sqrt((X2-X3)^2+(Y2-Y3)^2+(Z2-Z3)^2)
d(C3-C4)=sqrt((X3-X4)^2+(Y3-Y4)^2+(Z3-Z4)^2)
d(C4-C5)=sqrt((X4-X5)^2+(Y4-Y5)^2+(Z4-Z5)^2)
d(C5-O5)=sqrt((X5-X6)^2+(Y5-Y6)^2+(Z5-Z6)^2)
d(O5-C1)=sqrt((X6-X1)^2+(Y6-Y1)^2+(Z6-Z1)^2)
```
</details>

**Comment by drsalonsogil** [comment]
Dear Alicecomma,

Thank you for your analysis and your time! I knew it was an issue with the way the z-matrix has to deal with the ring. I agree that one bond distance is not defined. I tried to work on it, but, I would have included a worse issue at the time the program adds the substitute groups. For this reason, I strongly recommend optimizing the structures with QM or a FF. 

Best,

Santiago

**Comment by drsalonsogil** [comment]
Dear Alicecomma,

Is it possible to implement Vina-Carb in this code instead of AD Vina?

Thanks,

Santiago

**Comment by Alicecomma** [comment]
Absolutely, I've written my own C# console app around Vina-Carb. If you're running your code on linux, MCDock barely needs any changes. You would want to be sure the atom names are correct (C1 through O5 need to exist). If you're running on windows then windows subsystem for linux with ubuntu also requires minimal changes. MonteCarbo already runs through WSL.
